scripts/Muon/GUI/Common/threading_manager.py

```diff
@@ -0,0 +1,282 @@
+    cancelled = Signal()
+
+
+class Worker(QtCore.QThread):
+    """
+    Worker thread. Inherits from QThread to handle worker thread setup, signals and wrap-up. Using QThread here
+    (instead of QRunnable) so we can cancel the thread.
+
+    fn : A function to evaluate
+    kwargs : keyword arguments to pass to fn
+
+    returns are via signals defined in WorkerSignals()
+    """
+
+    def __init__(self, fn, **kwargs):
+        """
+        :param fn: The function to be evaluated.
+        :param kwargs: Keyword argument lists to pass to the function (e.g. x =[1,2,3] for f(x) where x is an integer).
+        """
+        super(Worker, self).__init__()
+        self.fn = threading_decorator(fn)
+        self.kwargs = kwargs
+        self.signals = WorkerSignals()
+        # Add callback so function can update its progress
+        self.kwargs['progress_callback'] = self.signals.progress
+
+    def run(self):
+        self.signals.started.emit()
+        try:
+            result, fails = self.fn(**self.kwargs)
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            for fail in fails:
+                # Exceptions caught whilst executing the function
+                exctype, value = type(fail), next(iter(fail.args), "")
+                self.signals.error.emit({"inputs": self.kwargs,
+                                         "exctype": exctype,
+                                         "value": value,
+                                         "traceback": traceback.format_exc()})
+        except:
+            # Catch errors not thrown by the function evaluation
+            exctype, value = sys.exc_info()[:2]  # traceback.print_exc()
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            self.signals.error.emit({"inputs": self.kwargs,
+                                     "exctype": exctype,
+                                     "value": value,
+                                     "traceback": traceback.format_exc()})
+        else:
+            # Successful evaluation
+            if 'progress_callback' in self.kwargs:
+                del self.kwargs['progress_callback']
+            self.signals.result.emit((self.kwargs, result))
+        finally:
+            self.signals.finished.emit()
+
+    def cancel(self):
+        self.signals.cancelled.emit()
+        # must use terminate (not quit/exit) as there is no event loop for the thread
+        self.terminate()
+
+
+class WorkerManager(QtWidgets.QWidget):
+    """
+    The WorkerManager class handles multi-threading of a function, by splitting an arbitrary number of list arguments
+    equally across a specified number of threads.
+
+    Example : for a function f(x,y) where x and y are ints, the WorkerManager allows for evaluation of
+    f(x=[1,2,3,4,5],y=[1,2,3,4,5]) by splitting the lists across the given number of threads. So for two threads the
```

Comment:
I like the flexibility of allowing arbitrary callables here. The syntax for the multi-argument functions feels a little clunky though as it looks like you have to split the lists up manually?
Could we just require the caller to pass the arguments as lists of tuples, for example: `f(x,y)` with args `((1,1), (2,2), (3,3), (4,4), (5,5))` naturally keeps the arguments together and I think makes it clearer what is being evaluated?

Reply:
At the moment I have limited arguments to keyword args, and I think that would prevent me doing this. However, I can add support for `*args*` as well, but it will make the `run` method of `Worker` more involved